### ISSUE TITLE
Struct field cannot have same name as type variable unless the type of the field is also the type variable

### ISSUE THREAD

**Issue by llbit** [JuliaLang/julia#46668]
I noticed that a struct like `struct S{T} T::String end` does not compile. Therefore I expected that `struct S{T} T::T end` would likewise not compile because both of them have a field with the same name as a type variable (the error message seems to imply there is a name collision causing the problem). However, the latter case was accepted by the Julia compiler.

This seems like a compiler error or an oversight in Julia semantics. Also the error message could be greatly improved to explain what the compiler thinks is wrong in the context of the surface syntax.

```
julia> VERSION
v"1.8.0"

julia> struct S{T} T::String end
ERROR: syntax: function argument and static parameter names must be distinct around REPL[2]:1
Stacktrace:
 [1] top-level scope
   @ REPL[2]:1

julia> struct S{T} T::T end

julia>
```

**Comment by vtjnash** [comment]
Incidentally fixed by #44664, since it removed the implied `where` clauses

**Comment by llbit** [comment]
I tested it again in `v"1.9.0-DEV.1301"` and the error is gone. Thank you!
